File: apps/ios/BitOS/DesignSystem/BrandQrCode.swift
import BusinessCore
import SwiftUI

/**
 * Branded QR (APP-022; legacy Flutter `BrandQrCode` parity): white card,
 * gray-900 modules, bitcoin-orange hex cover with the bolt glyph. The
 * matrix comes from the shared pure encoder through the bridge —
 * deterministic and identical to Android.
 */
struct BrandQrCodeView: View {
    let value: String
    var size: CGFloat = 224

    var body: some View {
        let kotlinRows = BusinessCoreBridge().qrMatrix(text: value)
        let rows: [Int64] = kotlinRows.map { $0.int64Value }
        if rows.isEmpty { return AnyView(EmptyView()) }
        let n = rows.count
        return AnyView(
            ZStack {
                RoundedRectangle(cornerRadius: 12)
                    .fill(Color.white)
                QrGridView(rows: rows, n: n)
                    .padding(10)
                BrandCover()
                    .frame(width: (size - 20) * 0.22, height: (size - 20) * 0.22)
            }
            .frame(width: size, height: size)
        )
    }
}

private struct QrGridView: View {
    let rows: [Int64]
    let n: Int

    var body: some View {
        Canvas { context, canvasSize in
            let cell = canvasSize.width / CGFloat(n + 1)
            let origin = cell / 2
            let moduleColor = Color(hex: 0x111827)
            for (y, rowBits) in rows.enumerated() {
                var bits = rowBits
                var x = 0
                while bits != 0 {
                    if bits & 1 != 0 {
                        let rect = CGRect(
                            x: origin + CGFloat(x) * cell,
                            y: origin + CGFloat(y) * cell,
                            width: cell + 0.5,
                            height: cell + 0.5
                        )
                        context.fill(Path(rect), with: .color(moduleColor))
                    }
                    bits >>= 1
                    x += 1
                }
            }
        }
    }
}

/// Orange hex disc with the white bolt glyph (scan-safe center cover).
private struct BrandCover: View {
    var body: some View {
        ZStack {
            Circle().fill(Color.white).padding(-6)
            HexShape()
                .fill(Color(hex: 0xF7931A))
            BoltGlyph()
                .fill(Color.white)
                .padding(10)
        }
    }
}

private struct BoltGlyph: Shape {
    func path(in rect: CGRect) -> Path {
        var p = Path()
        let w = rect.width
        let h = rect.height
        p.move(to: CGPoint(x: rect.minX + w * 0.60, y: rect.minY + h * 0.10))
        p.addLine(to: CGPoint(x: rect.minX + w * 0.25, y: rect.minY + h * 0.55))
        p.addLine(to: CGPoint(x: rect.minX + w * 0.52, y: rect.minY + h * 0.55))
        p.addLine(to: CGPoint(x: rect.minX + w * 0.40, y: rect.minY + h * 0.90))
        p.addLine(to: CGPoint(x: rect.minX + w * 0.75, y: rect.minY + h * 0.45))
        p.addLine(to: CGPoint(x: rect.minX + w * 0.48, y: rect.minY + h * 0.45))
        p.closeSubpath()
        return p
    }
}

File: apps/android/src/main/kotlin/space/bitos/app/ui/components/BrandQrCode.kt
package space.bitos.app.ui.components

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import space.bitos.core.nostr.QrCode

/**
 * Branded QR (APP-022; legacy Flutter `BrandQrCode` parity): white card,
 * gray-900 modules, bitcoin-orange hex cover with the bolt glyph in the
 * center. Matrix comes from the shared pure encoder — deterministic and
 * identical on both platforms.
 */
@Composable
fun BrandQrCode(
    value: String,
    sizeDp: Int = 224,
    contentDescription: String? = null,
) {
    val matrix = remember(value) { QrCode.encode(value) }
    if (matrix == null || matrix.isEmpty()) return
    val moduleColor = Color(0xFF111827)
    val coverColor = Color(0xFFF7931A)

    Box(
        modifier = Modifier
            .size(sizeDp.dp)
            .clip(RoundedCornerShape(12.dp))
            .background(Color.White)
            .padding(10.dp),
        contentAlignment = Alignment.Center,
    ) {
        Canvas(Modifier.size((sizeDp - 20).dp)) {
            val n = matrix.size
            val cell = this.size.width / (n + 1f) // + quiet-zone ring
            val origin = Offset(cell / 2, cell / 2)
            for (y in 0 until n) {
                val row = matrix[y]
                for (x in 0 until n) {
                    if (row[x]) {
                        drawRect(
                            color = moduleColor,
                            topLeft = Offset(origin.x + x * cell, origin.y + y * cell),
                            size = Size(cell + 0.5f, cell + 0.5f), // hairline overlap
                        )
                    }
                }
            }
        }
        // Brand cover: orange hex disc with the bolt glyph (scan-safe: a
        // quiet-zone-sized center keeps the code readable).
        Canvas(Modifier.size(((sizeDp - 20) * 0.22f).dp)) {
            val r = size.minDimension / 2f
            // White ring then orange hex.
            drawCircle(Color.White, radius = r * 1.15f)
            val hex = HexPath(center = Offset(r, r), radius = r)
            drawPath(hex, coverColor)
            // Bolt: simple polygon glyph.
            val bolt = androidx.compose.ui.graphics.Path().apply {
                moveTo(r + r * 0.10f, r - r * 0.55f)
                lineTo(r - r * 0.35f, r + r * 0.08f)
                lineTo(r + r * 0.02f, r + r * 0.08f)
                lineTo(r - r * 0.10f, r + r * 0.55f)
                lineTo(r + r * 0.35f, r - r * 0.08f)
                lineTo(r - r * 0.02f, r - r * 0.08f)
                close()
            }
            drawPath(bolt, Color.White)
        }
    }
}

/** Regular hexagon path (flat-top), matching HexShape geometry. */
private fun HexPath(center: Offset, radius: Float) = androidx.compose.ui.graphics.Path().apply {
    for (i in 0 until 6) {
        val angle = Math.PI / 6.0 + i * Math.PI / 3.0
        val x = center.x + (radius * kotlin.math.cos(angle)).toFloat()
        val y = center.y + (radius * kotlin.math.sin(angle)).toFloat()
        if (i == 0) moveTo(x, y) else lineTo(x, y)
    }
    close()
}

